packages/eas-cli/src/build/android/version.ts

```diff
@@ -1,50 +1,41 @@
 import { ExpoConfig } from '@expo/config';
 import { AndroidConfig } from '@expo/config-plugins';
 import { Platform, Workflow } from '@expo/eas-build-job';
-import fs from 'fs-extra';
+import { BuildProfile } from '@expo/eas-json';
 
+import {
+  getAppBuildGradleAsync,
+  parseGradleCommand,
+  resolveConfigValue,
+} from '../../project/android/gradleUtils';
 import { resolveWorkflowAsync } from '../../project/workflow';
 
-export async function readVersionCodeAsync(
+export async function maybeResolveVersionsAsync(
   projectDir: string,
-  exp: ExpoConfig
-): Promise<number | undefined> {
+  exp: ExpoConfig,
+  buildProfile: BuildProfile<Platform.ANDROID>
+): Promise<{ appVersion?: string; appBuildVersion?: string }> {
   const workflow = await resolveWorkflowAsync(projectDir, Platform.ANDROID);
   if (workflow === Workflow.GENERIC) {
-    const buildGradle = readBuildGradle(projectDir);
-    const matchResult = buildGradle?.match(/versionCode (.*)/);
-    if (matchResult) {
-      return Number(matchResult[1]);
-    } else {
-      return undefined;
-    }
-  } else {
-    return AndroidConfig.Version.getVersionCode(exp) ?? 1;
-  }
-}
+    const buildGradle = await getAppBuildGradleAsync(projectDir);
+    try {
+      const parsedGradleCommand = buildProfile.gradleCommand
+        ? parseGradleCommand(buildProfile.gradleCommand, buildGradle)
+        : undefined;
 
-export async function readVersionNameAsync(
-  projectDir: string,
-  exp: ExpoConfig
-): Promise<string | undefined> {
-  const workflow = await resolveWorkflowAsync(projectDir, Platform.ANDROID);
-  if (workflow === Workflow.GENERIC) {
-    const buildGradle = readBuildGradle(projectDir);
-    const matchResult = buildGradle?.match(/versionName ["'](.*)["']/);
-    if (matchResult) {
-      return matchResult[1];
-    } else {
-      return undefined;
+      return {
+        appVersion:
+          resolveConfigValue(buildGradle, 'versionName', parsedGradleCommand?.flavor) ?? '1.0.0',
+        appBuildVersion:
+          resolveConfigValue(buildGradle, 'versionCode', parsedGradleCommand?.flavor) ?? '1',
+      };
+    } catch (err: any) {
+      return {};
     }
   } else {
-    return exp.version;
-  }
-}
-
-function readBuildGradle(projectDir: string): string | undefined {
-  const buildGradlePath = AndroidConfig.Paths.getAppBuildGradleFilePath(projectDir);
-  if (!fs.pathExistsSync(buildGradlePath)) {
-    return undefined;
+    return {
+      appBuildVersion: String(AndroidConfig.Version.getVersionCode(exp) ?? 1),
```

Comment:
```suggestion
      appBuildVersion: String(AndroidConfig.Version.getVersionCode(exp)),
```
`getVersionCode` never returns falsy values. 
https://github.com/expo/expo-cli/commit/e3dab0c9fc917c490cf086c0cbe0fbdb950dbc33